File: src/main/java/net/netty/ServerHandler.java
package net.netty;

import client.Client;
import constants.ServerConstants;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import net.maple.PacketProcessor;
import net.maple.RecvOpcode;
import net.maple.handlers.PacketHandler;
import net.maple.packets.ConnectionPackets;
import util.crypto.MapleAESOFB;
import util.packet.Packet;
import util.packet.PacketReader;

import java.util.Arrays;

public class ServerHandler extends ChannelInboundHandlerAdapter {

    PacketProcessor processor = PacketProcessor.Companion.getInstance();
    int[] ignoreOps = {
            RecvOpcode.PONG.getValue(),
            RecvOpcode.UPDATE_SCREEN_SETTING.getValue(),
            RecvOpcode.USER_MOVE.getValue(),
            RecvOpcode.NPC_MOVE.getValue(),
            RecvOpcode.USER_CHANGE_STAT_REQUEST.getValue(),
            RecvOpcode.MOB_MOVE.getValue(),
            RecvOpcode.MOB_APPLY_CTRL.getValue(),
            RecvOpcode.ADMIN.getValue()
    };

    @Override
    public void channelActive(ChannelHandlerContext chc) {
        Channel channel = chc.channel();
        byte[] ivSend = {82, 48, 120, (byte) (Math.random() * 255)};
        byte[] ivRecv = {70, 114, 122, (byte) (Math.random() * 255)};

        Client client = new Client(channel, ivSend, ivRecv);

        client.write(ConnectionPackets.sendHello(ivSend, ivRecv, false));

        channel.attr(Client.Companion.getCLIENT_KEY()).set(client);
        channel.attr(Client.Companion.getCRYPTO_KEY()).set(new MapleAESOFB());

        System.out.printf("Opened session with %s%n", client.getIp());

        client.startPing();
    }

    @Override
    public void channelRead(ChannelHandlerContext chc, Object msg) {
        Packet packet = (Packet) msg;
        Channel channel = chc.channel();

        Client client = (Client) channel.attr(Client.Companion.getCLIENT_KEY()).get();
        PacketReader packetReader = client.getReader().next(packet);

        short opCode = packetReader.readShort();
        final PacketHandler packetHandler = processor.getHandler(opCode);

        String hex = Integer.toHexString(opCode);
        if (ServerConstants.LOG && Arrays.stream(ignoreOps).noneMatch(p -> p == opCode)) {
            if (packetHandler == null) {
                System.out.println("[RECEIVED] packet " + opCode + " (" + (hex.length() == 1 ? "0x0" : "0x") + hex.toUpperCase() + ")");
            } else {
                String className = packetHandler.getClass().getName();
                if (!className.contains("DoNothingHandler")) {
                    System.out.println("[RECEIVED] " + className);
                }
            }
            //System.out.printf("data: %s.%n", packet.toString());
        }

        if (packetHandler != null) {
            if (packetHandler.validateState(client)) {
                packetHandler.handlePacket(packetReader, client);
            } else {
                System.out.printf("Client failed to validate state for packet %s.%n", opCode);
                channel.close();
            }
        } else {
            System.out.printf("Received completely unhandled packet %s.%n", packet.toString());
        }
    }

    @Override
    public void channelInactive(ChannelHandlerContext chc) {
        Channel ch = chc.channel();

        Client c = (Client) ch.attr(Client.Companion.getCLIENT_KEY()).get();

        c.disconnect(); // todo npe on character? (lateinit, dc before init)
        //c.softDisconnect(c.isLoggedIn()); // handle this is we don't soft disconnect through handler

        c.cancelPingTask();

        // remove after debug stage
        System.out.printf("[Debug] Closed session with %s.%n", c.getIp());
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext chc, Throwable cause) {
        // Close the connection when an exception is raised.
        Channel ch = chc.channel();
        Client c = (Client) ch.attr(Client.Companion.getCLIENT_KEY()).get();
        //c.disconnect();
        //System.out.println(cause);
        cause.printStackTrace();
        //chc.close();
    }
}

File: src/main/java/net/netty/ServerHandler.kt
package net.netty

import client.Client
import constants.ServerConstants
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import net.maple.PacketProcessor
import net.maple.RecvOpcode
import net.maple.packets.ConnectionPackets
import net.netty.NettyClient.Companion.CLIENT_KEY
import net.netty.NettyClient.Companion.CRYPTO_KEY
import util.crypto.MapleAESOFB
import util.packet.Packet

class ServerHandler(val port: Int) : ChannelInboundHandlerAdapter() {

    private var processor = PacketProcessor
    private var ignoreOps = intArrayOf(
            RecvOpcode.PONG.value,
            RecvOpcode.UPDATE_SCREEN_SETTING.value,
            RecvOpcode.USER_MOVE.value,
            RecvOpcode.NPC_MOVE.value,
            RecvOpcode.USER_CHANGE_STAT_REQUEST.value,
            RecvOpcode.MOB_MOVE.value,
            RecvOpcode.MOB_APPLY_CTRL.value,
            RecvOpcode.ADMIN.value,
            RecvOpcode.PET_MOVE.value
    )

    override fun channelActive(chc: ChannelHandlerContext) {
        val channel = chc.channel()
        val ivSend = byteArrayOf(82, 48, 120, (Math.random() * 255).toInt().toByte())
        val ivRecv = byteArrayOf(70, 114, 122, (Math.random() * 255).toInt().toByte())

        val client = Client(channel, ivSend, ivRecv)

        client.write(ConnectionPackets.sendHello(ivSend, ivRecv, false))

        channel.attr(CLIENT_KEY).set(client)
        channel.attr(CRYPTO_KEY).set(MapleAESOFB())

        System.out.printf("Opened session on port $port with %s%n", client.ip)

        client.startPing()
    }

    override fun channelRead(chc: ChannelHandlerContext, msg: Any) {
        val packet = msg as Packet
        val channel = chc.channel()

        val client = channel.attr(CLIENT_KEY).get() as Client
        val packetReader = client.reader.next(packet)

        val opCode = packetReader.readShort()
        val packetHandler = processor.getHandler(opCode)

        val hex = Integer.toHexString(opCode.toInt())
        if (ServerConstants.LOG && !ignoreOps.contains(opCode.toInt())/* Arrays.stream(ignoreOps).noneMatch { it == opCode.toInt() }*/) {
            if (packetHandler == null) {
                println("[RECEIVED] packet " + opCode + " (" + (if (hex.length == 1) "0x0" else "0x") + hex.toUpperCase() + ")")
            } else {
                val className = packetHandler.javaClass.name
                if (!className.contains("DoNothingHandler")) {
                    println("[RECEIVED] $className")
                }
            }
            //System.out.printf("data: %s.%n", packet.toString());
        }

        packetHandler?.let {
            if (it.validateState(client)) {
                it.handlePacket(packetReader, client)
            } else {
                System.out.printf("Client failed to validate state for packet %s.%n", opCode)
                channel.close()
            }
        } ?: System.out.printf("Received completely unhandled packet %s.%n", packet.toString())
    }

    override fun channelInactive(chc: ChannelHandlerContext) {
        val ch = chc.channel()
        val c = ch.attr(CLIENT_KEY).get() as Client

        if (c.isInitialized()) {
            c.disconnect()
        }
        c.cancelPingTask()

        // remove after debug stage
        System.out.printf("[Debug] Closed session on port $port with %s.%n", c.ip)
    }

    override fun exceptionCaught(chc: ChannelHandlerContext, cause: Throwable) {
        // Close the connection when an exception is raised.
        val ch = chc.channel()
        val c = ch.attr(CLIENT_KEY).get() as Client
        //c.disconnect();
        //System.out.println(cause);
        System.err.println("[exceptionCaught]")
        cause.printStackTrace()
        //chc.close();
    }
}